src/Npgsql/NpgsqlConnection.cs

```diff
@@ -501,16 +501,17 @@ public new NpgsqlTransaction BeginTransaction(IsolationLevel level)
         /// <summary>
         /// Enlist transation.
         /// </summary>
-        public override void EnlistTransaction(Transaction transaction)
+        public override void EnlistTransaction([CanBeNull]Transaction transaction)
         {
-            if (EnlistedTransaction != null)
+            var enlistedTransaction = EnlistedTransaction;
+            if (enlistedTransaction != null)
             {
-                if (EnlistedTransaction.Equals(transaction))
+                if (enlistedTransaction.Equals(transaction))
                     return;
                 try
                 {
-                    if (EnlistedTransaction.TransactionInformation.Status == System.Transactions.TransactionStatus.Active)
-                        throw new InvalidOperationException($"Already enlisted to transaction (localid={EnlistedTransaction.TransactionInformation.LocalIdentifier})");
+                    if (enlistedTransaction.TransactionInformation.Status == System.Transactions.TransactionStatus.Active)
+                        throw new InvalidOperationException($"Already enlisted to transaction (localid={enlistedTransaction.TransactionInformation.LocalIdentifier})");
```

Comment:
New bug discovered: after scope disposal in distributed case, the transaction status may still be `Active`! This happens when the supplied transaction was not the original one but a `.Clone()` of it... Illustrated in this NHibernate [test case](https://github.com/nhibernate/nhibernate-core/pull/627/files#r125465830). Waiting a bit allows to get the correct status...

I do not think we can do much against that later case. The main usual case (enlisting with the original transaction, not with a clone) will at least be protected by that code.